src/openassetio-python/tests/cmodule/test_errors.py

```diff
@@ -66,3 +70,82 @@ def test_when_ConfigurationException_thrown_then_can_be_caught_as_InputValidatio
         )
         with pytest.raises(errors.InputValidationException, match="Explosion!"):
             _openassetio_test.throwException(errors.ConfigurationException.__name__, "Explosion!")
+
+    @pytest.mark.parametrize("exception_type", all_exceptions)
+    def test_when_python_exception_thrown_then_can_be_caught_in_cpp(
+        self, exception_type, exception_thrower
+    ):
+        exception = make_exception(exception_type)
+
+        exception_thrower.callee.side_effect = exception
+
+        assert _openassetio_test.isPythonExceptionCatchableAs(
+            exception_thrower, exception_type.__name__
+        )
+
+    @pytest.mark.parametrize("exception_type", all_exceptions)
+    def test_when_python_exception_thrown_then_can_be_caught_in_cpp_rethrown_and_caught_in_python(
+        self, exception_type, exception_thrower
+    ):
+        exception = make_exception(exception_type)
+
+        exception_thrower.callee.side_effect = exception
+
+        with pytest.raises(exception_type) as err:
+            _openassetio_test.throwPythonExceptionCatchAsCppExceptionAndRethrow(
+                exception_thrower, exception_type.__name__
+            )
+
+        assert err.value is exception
+
+    def test_when_non_OpenAssetIO_exception_thrown_then_not_translated_to_cpp_exception(
+        self, exception_thrower
+    ):
+        # Same name, but different exception.
+        class OpenAssetIOException(Exception):
+            pass
+
+        exception = OpenAssetIOException()
+
+        exception_thrower.callee.side_effect = exception
+
+        with pytest.raises(OpenAssetIOException):
+            # Should fail to catch and propagate back.
+            _openassetio_test.isPythonExceptionCatchableAs(
+                exception_thrower, "OpenAssetIOException"
+            )
+
+
+def make_exception(exception_type):
+    if exception_type == errors.BatchElementException:
+        error = errors.BatchElementError(
+            errors.BatchElementError.ErrorCode.kInvalidEntityReference, "Explosion!"
+        )
+        exception = exception_type(123, error, "Another explosion!")
+    else:
+        exception = exception_type("Explosion!")
+
+    return exception
+
+
+class PyExceptionThrower(_openassetio_test.ExceptionThrower):
```

Comment:
[Consider] This is really well documented in c++ - might be worth a cursory note here on what this is doing and how.

Reply:
Added a docstring in 9103b05c